### PR TITLE
Issues #17 and #18

### PR DESCRIPTION
**Pull request by jim-works**
This branch contains the implementation and tests for issues #17 and #18

### CODE REVIEW

**Comment by joshmackwilliams** [comment]
That last commit is to resolve a trivial merge conflict in .gitignore. I will actually review this tomorrow before the meeting.

File: src/planning/path_planner/design/path_planner-design.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Path planner {#path-planner}
```
Comment (wheitman) [inline src/planning/path_planner/design/path_planner-design.md:1]: Love the design doc, thanks for doing this.

File: src/planning/path_planner/include/path_planner/map_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef PATH_PLANNER_MAP_UTILS_HPP
+#define PATH_PLANNER_MAP_UTILS_HPP
+
+
+#include "autoware_auto_msgs/msg/trajectory.hpp"
+#include <lanelet2_core/geometry/Lanelet.h>
+#include <lanelet2_core/LaneletMap.h>
+#include <lanelet2_core/primitives/Point.h>
+#include <lanelet2_core/geometry/LineString.h>
+#include <autoware_auto_msgs/msg/trajectory.hpp>
```
Comment (wheitman) [inline src/planning/path_planner/include/path_planner/map_utils.hpp:19]: We should transition to using `voltron_msgs` in the future, migrating and modifying Autoware msg files as necessary. This gives us greater clarity and control over our message formats. And I suspect that Autoware's message approach can be simplified for our uses...

File: src/planning/path_planner/src/parameterized_spline.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
+// Copyright 2020 The Autoware Foundation
```
Comment (wheitman) [inline src/planning/path_planner/src/parameterized_spline.cpp:1]: Is this Autoware code? I'm not seeing it in their repo.

File: src/planning/path_planner/src/path_planner.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+}
+autoware_auto_msgs::msg::TrajectoryPoint convertToTrajectoryPoint(const lanelet::ConstPoint3d &pt)
+{
+    autoware_auto_msgs::msg::TrajectoryPoint trajectory_point;
+    trajectory_point.x = static_cast<float32_t>(pt.x());
+    trajectory_point.y = static_cast<float32_t>(pt.y());
+    trajectory_point.longitudinal_velocity_mps = 0;
+    return trajectory_point;
+}
+//resolution should be < 0.1 to get good approximation
```
Comment (wheitman) [inline src/planning/path_planner/src/path_planner.cpp:75]: 0.1 what? Meters? You used "units" earlier, but the Lanelet data is stored in meters, right?

File: src/planning/path_planner/include/path_planner/spline.hpp [hunk 5]
```diff
@@ -0,0 +1,751 @@
...
+ *  modify it under the terms of the GNU General Public License
+ *  as published by the Free Software Foundation; either version 2
+ *  of the License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
```
Comment (wheitman) [inline src/planning/path_planner/include/path_planner/spline.hpp:20]: @jim-works would you include the [GPL v3 license text](https://www.gnu.org/licenses/gpl-3.0.txt) at [the end of our LICENSE file](https://github.com/Nova-UTD/navigator/blob/360ab836156313d2ffd67ab94012e24a8d8c25ba/LICENSE#L226), along with an explanatory note of which part of our code uses the file, e.g. "Tino Kluge's cubic spline library", including a link to their site (https://kluge.in-chemnitz.de/opensource/spline/)

**Review by wheitman — CHANGES_REQUESTED** [review]
Great work, Jim. Please make these quick fixes as soon as you're able. @joshmackwilliams, please add any comments. Thanks!

**Review by joshmackwilliams — APPROVED** [review]
Looks good to me

File: src/planning/path_planner/include/path_planner/map_utils.hpp (hunk 2 above)
Comment (jim-works) [reply src/planning/path_planner/include/path_planner/map_utils.hpp:19]: Sounds good

File: src/planning/path_planner/src/parameterized_spline.cpp (hunk 3 above)
Comment (jim-works) [reply src/planning/path_planner/src/parameterized_spline.cpp:1]: no, I initially copied their class and forgot to remove this comment, oops

File: src/planning/path_planner/src/path_planner.cpp (hunk 4 above)
Comment (jim-works) [reply src/planning/path_planner/src/path_planner.cpp:75]: It should be 0.1 of whatever units we decide to use, so meters is fine. It's more of a note to myself, for the future from my testing earlier. Will be more robust later.

File: src/planning/path_planner/include/path_planner/spline.hpp (hunk 5 above)
Comment (jim-works) [reply src/planning/path_planner/include/path_planner/spline.hpp:20]: Sure

File: src/planning/path_planner/test/test_path_planner.cpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  lanelet::LineString3d ls3(lanelet::utils::getId(), center_points);
+
+  lanelet::Lanelet ll(id, ls1, ls2);
+  ll.setCenterline(ls3);
+  ll.setAttribute(lanelet::AttributeName::SpeedLimit, velocity);
+
+  return lanelet::utils::createMap({ll});
+}
+
+HADMapRoute getARoute(const int64_t lane_id, const float32_t length)
```
[the remark this answers is not in the document]
Comment (jim-works) [reply src/planning/path_planner/test/test_path_planner.cpp:50]: Yep

**Comment by wheitman** [comment]
@jim-works please add that license text and function description, then we can merge. Thanks 👍

**Comment by jim-works** [comment]
Apologies that this took so long, school was killing me this week. Also, I did not change to voltron_msgs, it didn't look like the package had the trajectory struct I was using (unless I am missing something?)

**Review by wheitman — APPROVED** [review]
Looks good, thanks Jim.

**Comment by wheitman** [comment]
> Apologies that this took so long, school was killing me this week. Also, I did not change to voltron_msgs, it didn't look like the package had the trajectory struct I was using (unless I am missing something?)

Right, I meant that more as a todo for future work. Moving to a unified message library under voltron_msgs will be a team-wide effort, but you don't have to worry about that at the moment.

**Pull request merged**